sympy/solvers/simplex.py

```diff
@@ -0,0 +1,966 @@
+
+
+def linprog(cd, A=None, b=None, A_eq=None, b_eq=None, bounds=None):
+    """Return the minimization of ``c*x - d`` with the given
+    constraints ``A*x <= b`` and ``A_eq*x = b_eq``. Unless bounds
+    are given, variables will have nonnegative values in the solution.
+
+    If ``cd`` is a single matrix then ``d`` will be zero; otherwise
+    it should be tuple with two matrices: ``c`` and ``d``.
+
+    By default, all variables will be nonnegative. If ``bounds``
+    is given as a single tuple, ``(lo, hi)``, then all variables
+    will be constrained to be between ``lo`` and ``hi``. Use
+    None for a ``lo`` or ``hi`` if it is unconstrained in the
+    negative or positive direction, respectively, e.g.
+    ``(None, 0)`` indicates nonpositive values. To set
+    individual ranges, pass a list with length equal to the
+    number of columns in ``A``, each element being a tuple; if
+    only a few variables take on non-default values they can be
+    passed as a dictionary with keys giving the corresponding
+    column to which the variable is assigned, e.g. ``bounds={2:
+    (1, 4)}`` would limit the 3rd variable to have a value in
+    range ``[1, 4]``.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.simplex import linprog
+    >>> from sympy import symbols, Eq, linear_eq_to_matrix as M, Matrix
+    >>> x = x1, x2, x3, x4 = symbols('x1:5')
+    >>> X = Matrix(x)
+    >>> c, d = cd = M(5*x2 + x3 + 4*x4 - x1, x)
+    >>> a, b = M([5*x2 + 2*x3 + 5*x4 - (x1 + 5)], x)
+    >>> aeq, beq = M([Eq(3*x2 + x4, 2), Eq(-x1 + x3 + 2*x4, 1)], x)
+    >>> constr = [i <= j for i,j in zip(a*X, b)]
+    >>> constr += [Eq(i, j) for i,j in zip(aeq*X, beq)]
+    >>> linprog(cd, a, b, aeq, beq)
+    (9/2, [0, 1/2, 0, 1/2])
+    >>> assert all(i.subs(dict(zip(x, _[1]))) for i in constr)
+
+    See Also
+    ========
+    lpmin, lpmax
+    """
+
+    # convert scipy-like matrices to expressions
+
+    ## the inequalities
+    if A is None:
+        if b is not None:
+            raise ValueError('A and b must both be given')
+        A, b = Matrix(0, 0, []), Matrix(0, 0, [])
+    else:
+        A, b = [Matrix(i) for i in (A, b)]
+
+    ## the equalities
+    if A_eq is None:
+        if not b_eq is None:
+            raise ValueError('A_eq and b_eq must both be given')
+    else:
+        A_eq, b_eq = [Matrix(i) for i in (A_eq, b_eq)]
+        # if x == y then x <= y and x >= y (-x <= -y)
+        A = A.col_join(A_eq)
+        A = A.col_join(-A_eq)
+        b = b.col_join(b_eq)
+        b = b.col_join(-b_eq)
+
+    ## the objective
+    if isinstance(cd, MatrixBase):
+        C, D = cd, Matrix([0])
+    elif len(cd) == 2 and all(isinstance(i, (MatrixBase, list)
+            ) for i in cd):
+        C, D = [Matrix(i) for i in cd]
+    elif isinstance(cd, list) and isinstance(cd[0], list) and len(
+            cd[0]) == A.cols:
+        C, D = Matrix(cd), Matrix([0])
+    else:
+        raise ValueError('expecting one or two matrices/lists for cd')
```

Comment:
Is it actually useful to allow these different possibilities? I find it hard to follow what this function does because of the ambiguity about its arguments.

Reply:
Especially for a public function we need to be careful about signature. I don't see a good reason to stray from what SciPy and Matlab do in this function signature.

Reply:
This is my only real concern for this PR.

Reply:
Anyone using scipy can (up to the bounds) pass the same variables there as here. The only difference is that this `linprog` allows a `D` -- but since that doesn't factor into the solution process we could leave it off and let the user ponder why it is not needed. For simplicity I am not opposed to omitting `D`.

Reply:
I should take a closer look to understand how these APIs work.

No objection from my side but I thought I understood this and probably I didn't...

Reply:
It has now been omitted.

Reply:
> No objection from my side but I thought I understood this and probably I didn't..

I am very appreciative of the time given for review. I agree that when working with symbolic constraints that we should let the user make the assumptions rather than imposing an assumption made by a support routine. So now `lpmin/max` are WYSIWYG in terms of constraints and infinities are disallowed (because those are part of the default assumption for variables). See the simple 1-D example in the OP.

The basic idea is that a linear function can be minimized on a simplex (n-dim linear region). `_simplex` does the task of minimizing `C*x - D` with `A*x <= B` for nonnegative variables, `x`. Everything else is scaffolding to get a system expressed with Le, Ge or Eq and desired bounds expressed in terms of A, B, C, D and to handle finding the max or the min.